### ISSUE TITLE
rebuild in a chroot: cannot pivot old root directory

### ISSUE THREAD

**Issue by martinetd** [NixOS/nix#5076]
### Describe the bug
nixos-rebuild cannot be used in a chroot to recover a system booting a usb device or similar.
I'm pretty sure that used to work.
```
[root@xxx:/etc/nixos]# nixos-rebuild build --show-trace
building the system configuration...
error: cannot pivot old root directory onto '/nix/store/5vrfx0qc9hllcq8harrskllr72rwangm-stage-2-init.sh.drv.chroot/real-root': Invalid argument

       … while setting up the build environment
error: getting status of '/tmp/nixos-rebuild.3FL7Nv/result': No such file or directory
```

on nix unstable (flake enabled), nixpkgs 16105403bdd843540cbef9c63fc0f16c1c6eaa70 (nixos-unstable)

### Steps To Reproduce
1. build a nixos OS that can be mounted separated from main system (which can be nixos), say it's mounted on /mnt - make sure it has things to rebuild as a noop rebuild will work
2. bind mount /proc, /sys, /dev and /dev/pts to new root
3. chroot /mnt /run/current-system/sw/bin/sh and fix PATH if required
4. nixos-rebuild build

### Expected behavior

I'd guess a way of setting up the build environment that doesn't rely on pivot_root if that failed?
I'd be happy with an option --continue-anyway or something ugly without sandbox at this point; looking at the code it does pivot_root just before chroot to provide extra protection? but it looks like it'd probably work without it in a bind.

I also noticed the pivot_root is only done if useChroot is set, which is basically on sandbox, but nixos-rebuild build --impure didn't help and I'm not sure how to disable sandbox for a single run of nixos-rebuild (couldn't figure how to pass nix options directly), I'd probably be happy with a clear error message that explains how to disable it as a workaround.

**`nix-env --version` output**

nix-env (Nix) 2.4pre20210707_02dd6bb

**Additional context**

FWIW it's the second time I've run into this, in both case it was more or less something broke and I wanted to take the opportunity to change partitions quite heavily and figured it'd be easier from a usb stick than from the running system.
Basically boot into USB (I have a sysrescuecd around), copy filesystems around, fix hardware-configuration.nix and nixos-rebuild should get me sorted... But it's no help if I can't run it.

I guess I'd have less trouble from a nixos boot disk used as with nixos-install but that's not always available (think datacenter provided netboot for example), and being able to fix the system from a chroot would be invaluable.

**Comment by grahamc** [comment]
I'm running `nixos-install` inside a systemd service and I'm seeing this same issue. I don't think I'm in a chroot, though, which makes it a bit mysterious as to why I'm seeing it.

**Comment by grahamc** [comment]
I chatted with @cleverca22 and heard passing `--option sandbox false` fixes it. I've tried that but it didn't work. This was on Nixpkgs rev `fe01052444c1d66ed6ef76df2af798c9769e9e79`. I'll be looking further shortly.

**Comment by grahamc** [comment]
In my case the issue stems from my `/` being a `rootfs`, and `man 2 pivot_root` says you can't use pivot_root on a rootfs. I ended up patching my kernel with the attached patch to make it work.
[PATCH-v2-1-1-mnt-add-support-for-non-rootfs-initramfs.txt](https://github.com/NixOS/nix/files/7162294/PATCH-v2-1-1-mnt-add-support-for-non-rootfs-initramfs.txt)

**Comment by martinetd** [comment]
FWIW `--option sandbox false` worked for me (nix 2.17.0), after which I could use nixos-rebuild boot from the chroot to try to fix things for next boot (switch wouldn't work because it couldn't talk to the host's systemd, obviously)

I have my config in a flake in a not so obvious way, but it boiled down to this:
```
# nix --option sandbox false -L build .#nixosConfigurations.myhostname.config.system.build.toplevel 
# nixos-rebuild boot
```

**Comment by M-Gregoire** [comment]
As part of fixing the OVH install script (https://github.com/nix-community/nixos-install-scripts/pull/23), I've encountered the same error. Initially with:
```
curl -L https://nixos.org/nix/install | sh -s -- --daemon --yes
```
Which got fixed by:
```
mkdir -p /etc/nix
echo "build-users-group =" > /etc/nix/nix.conf
echo "sandbox = false" >> /etc/nix/nix.conf

# https://github.com/NixOS/nix/issues/7790#issuecomment-1451990482
mount --bind / /
```

However, a bit more surprisingly, I now encounter the error with `nix-channel --update`
```
-bash-5.0# nix-channel --list
nixpkgs https://nixos.org/channels/nixos-24.05
-bash-5.0# nix-channel --update
unpacking 1 channels...
error:
       … while setting up the build environment

       error: cannot pivot old root directory onto '/nix/store/dy8qb5adj97cfmmf4ah5p07gbs0x1hib-nixpkgs-24.05.drv.chroot/root/real-root': Invalid argument
error: program '/nix/store/hdy82qidsybc3fg561pqfwagv44vschb-nix-2.24.10/bin/nix-env' failed with exit code 1
```

Do you have any idea what could be the issue here?
